src/marked.js

```diff
@@ -139,44 +139,84 @@ marked.defaults = defaults;
  */
 
 marked.use = function(extension) {
-  const opts = merge({}, extension);
-  if (extension.renderer) {
-    const renderer = marked.defaults.renderer || new Renderer();
-    for (const prop in extension.renderer) {
-      const prevRenderer = renderer[prop];
-      renderer[prop] = (...args) => {
-        let ret = extension.renderer[prop].apply(renderer, args);
-        if (ret === false) {
-          ret = prevRenderer.apply(renderer, args);
+  if (!Array.isArray(extension)) { // Wrap in array if not already to unify processing
+    extension = [extension];
+  }
+  const opts = merge({}, ...extension);
+  opts.tokenizer = null;
+  opts.renderer = null;
+  const extensions = {};
+
+  extension.forEach((ext) => {
+    //= =-- Parse "addon" extensions --==//
+    if (ext.renderer && ext.name) { // Renderers must have 'name' property
+      extensions[ext.name] = ext.renderer;
+    }
+    if (ext.tokenizer && ext.level) { // Tokenizers must have 'level' property
+      if (extensions[ext.level]) {
+        extensions[ext.level].push(ext.tokenizer);
+      } else {
+        extensions[ext.level] = [ext.tokenizer];
+      }
+      if (ext.start) { // Regex to check for start of token
+        if (ext.level === 'block') {
+          extensions.startBlock = extensions.startBlock
+            ? new RegExp(extensions.startBlock.source + '|' + ext.start.source)
+            : ext.start;
+        } else if (ext.level === 'inline') {
+          extensions.startInline = extensions.startInline
+            ? new RegExp(extensions.startInline.source + '|' + ext.start.source)
+            : ext.start;
         }
-        return ret;
-      };
+      }
     }
-    opts.renderer = renderer;
-  }
-  if (extension.tokenizer) {
-    const tokenizer = marked.defaults.tokenizer || new Tokenizer();
-    for (const prop in extension.tokenizer) {
-      const prevTokenizer = tokenizer[prop];
-      tokenizer[prop] = (...args) => {
-        let ret = extension.tokenizer[prop].apply(tokenizer, args);
-        if (ret === false) {
-          ret = prevTokenizer.apply(tokenizer, args);
+
+    //= =-- Parse "overwrite" extensions --==//
+    if (ext.renderer && !ext.name) {
+      const renderer = marked.defaults.renderer || new Renderer();
+      for (const prop in ext.renderer) {
+        const prevRenderer = renderer[prop];
+        renderer[prop] = (...args) => {
+          let ret = ext.renderer[prop].apply(renderer, args);
+          if (ret === false) {
+            ret = prevRenderer.apply(renderer, args);
```

Comment:
if `ext.renderer` is a new renderer than `prevRenderer` will be `undefined` and this will fail.

We should make sure there is a unit test for a new renderer returning false. It should probably just render `''`.

Reply:
same for the tokenizer

Reply:
This should be blocked by the if statement above, right? If we have a new renderer without a `name` property, then we don't even enter this section.

https://github.com/markedjs/marked/blob/7bf58eb5e218fcf75964cd02d952914b85625093/src/marked.js#L175

https://github.com/markedjs/marked/blob/7bf58eb5e218fcf75964cd02d952914b85625093/src/marked.js#L189

Reply:
Added a unit test with a new renderer that returns false. The parser will use `''` if the return value from the renderer is falsy.